### PR TITLE
[Mailer] maintain sender/recipient name in SMTP envelopes

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #33383
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by xabbuh** [comment]
The build failure is unrelated to the changes.

**Comment by xabbuh** [comment]
I found some more places. Ready to be reviewed now.

**Comment by fbourigault** [comment]
As `Address` is an immutable object, it should be better to use those as-is instead of creating new instances.

Moreover, we are lacking some escaping to flawlessly use addresses names in email headers.

AFAIK, commas and double quotes have to be escaped.

I use the following code in my project which ATM use the official postmark client:
```php
        $name = preg_replace('/"/u', '\"', $name);
        if (preg_match('/,/u', $name)) {
            $name = sprintf('"%s"', $name);
        }
```

File: src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php [hunk 1]
```diff
@@ -73,7 +73,7 @@ private static function getRecipientsFromHeaders(Headers $headers): array
         foreach (['to', 'cc', 'bcc'] as $name) {
             foreach ($headers->all($name) as $header) {
                 foreach ($header->getAddresses() as $address) {
```
Comment (fbourigault) [inline src/Symfony/Component/Mailer/DelayedSmtpEnvelope.php]: Is it possible to drop one nested loop by using `$recipients = array_merge($recipients, $header->getAddresses()`?

File: src/Symfony/Component/Mailer/SmtpEnvelope.php [hunk 2]
```diff
@@ -39,7 +39,7 @@ public static function create(RawMessage $message): self
 
     public function setSender(Address $sender): void
     {
-        $this->sender = new Address($sender->getAddress());
+        $this->sender = $sender;
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/SmtpEnvelope.php]: This was done this way as the Sender (`MAIL FROM` in the SMTP protocol) only accepts an address without the "name" part. But indeed, now that we are also using it for HTTP/API, we need to keep the full address. You must fix `$this->doMailFromCommand($envelope->getSender()->toString());` in `SmtpTransport` though to only use the address.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
I think this one needs to be rebased on 4.3 as this is a bug fix, right? @xabbuh Can you rebase?

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**